hikari/impl/event_manager.py

```diff
@@ -113,82 +136,168 @@ async def on_channel_delete(self, shard: gateway_shard.GatewayShard, payload: da
+            emojis = event.emojis
+            guild = event.guild
+            guild_id = guild.id
+            members = event.members
+            presences = event.presences
+            roles = event.roles
+            voice_states = event.voice_states
 
-            self._cache.clear_voice_states_for_guild(event.guild.id)
-            for voice_state in event.voice_states.values():
-                self._cache.set_voice_state(voice_state)
-
-            members_declared = self._app.intents & intents_.Intents.GUILD_MEMBERS
-            presences_declared = self._app.intents & intents_.Intents.GUILD_PRESENCES
+        else:
+            event = None
+            channels = None
+            emojis = None
+            guild = None
+            guild_id = snowflakes.Snowflake(payload["id"])
+            members = None
+            presences = None
+            roles = None
+            voice_states = None
 
-            # When intents are enabled discord will only send other member objects on the guild create
-            # payload if presence intents are also declared, so if this isn't the case then we also want
-            # to chunk small guilds.
-            if members_declared and (event.guild.is_large or not presences_declared):
-                # We create a task here instead of awaiting the result to avoid any rate-limits from delaying dispatch.
-                nonce = f"{shard.id}.{_fixed_size_nonce()}"
+        if self._cache:
+            if guild:
+                self._cache.update_guild(guild)
+
+            if channels:
+                self._cache.clear_guild_channels_for_guild(guild_id)
+                for channel in channels.values():
+                    self._cache.set_guild_channel(channel)
+
+            if emojis:
+                self._cache.clear_emojis_for_guild(guild_id)
+                for emoji in emojis.values():
+                    self._cache.set_emoji(emoji)
+
+            if roles:
+                self._cache.clear_roles_for_guild(guild_id)
+                for role in roles.values():
+                    self._cache.set_role(role)
+
+            if members:
+                # TODO: do we really want to invalidate these all after an outage.
+                self._cache.clear_members_for_guild(guild_id)
+                for member in members.values():
+                    self._cache.set_member(member)
+
+            if presences:
+                self._cache.clear_presences_for_guild(guild_id)
+                for presence in presences.values():
+                    self._cache.set_presence(presence)
+
+            if voice_states:
+                self._cache.clear_voice_states_for_guild(guild_id)
+                for voice_state in voice_states.values():
+                    self._cache.set_voice_state(voice_state)
+
+        recv_chunks = self._enabled_for_event(shard_events.MemberChunkEvent) or self._cache_enabled_for(
+            config.CacheComponents.MEMBERS
+        )
```

Comment:
What if we spin up the MemberChunkEvent listener when we receive the GuildAvailable?

Reply:
What would the benefit be to that

Reply:
I mean, what if the users register the event listener when they receive the `GuildAvailableEvent`. This will cause it to not be registered before doing this check and cause the event to never be dispatched

*Really should have worded that correctly, sorry for that*

Reply:
Well subscribing an event listener wouldn't yield to the event loop and neither does anything in the consumer functions so that race condition would only come around if they're subscribing the listener from another thread and this already isnt documented as being thread safe so...